### DISCUSSION TITLE
Draw filled triangle without using convex_polygon?

### DISCUSSION THREAD

**Discussion by cadennabors** [emilk/egui#3495] [Q&A] [answered]
How can I draw a filled in triangle using the painter without using convex polygon? 

I can't use convex_polygon as it has a ton of issues when drawing sharp edges.
For example, when drawing using lines only, I get this:
![image](https://github.com/emilk/egui/assets/78510667/344aa27a-7752-4cb1-9444-3ff3c9531220)

But when I use convex_polygon with transparent fill, I should get the same as before, but instead I get this:
![image](https://github.com/emilk/egui/assets/78510667/63ff1d35-495a-4567-a566-00832cddcf36)

According to @emilk in a 2022 issue about convex_polygon anti-aliasing issues, 

"I pushed a few fixes now. The last fix comes at the cost of having sharp edges extrude weirdly, and extrude further the sharper they are."

Are there any alternatives that will let me accomplish this?

**Comment by JumpyLionnn** [comment, marked as answer]
If you want to draw filled triangles you can use the `Mesh` api https://docs.rs/egui/latest/egui/struct.Mesh.html
You can add the vertices you need and then connect them as triangles with indices.
For example to draw a black triangle within a rect you can use the following code:
```rust
let mut mesh = egui::Mesh::default();
mesh.colored_vertex(rect.center_top(), egui::Color32::BLACK);
mesh.colored_vertex(rect.left_bottom(), egui::Color32::BLACK);
mesh.colored_vertex(rect.right_bottom(), egui::Color32::BLACK);
mesh.add_triangle(0, 1, 2);
ui.painter().add(egui::Shape::mesh(mesh));
```

**Reply by cadennabors** [reply]
Thank you!
